### PR TITLE
csrf: init implementation

### PR DESCRIPTION
**Pull request by dsh0416**

### CODE REVIEW

File: csrf.go [hunk 1]
```diff
@@ -1 +1,295 @@
...
+	// Return cookie path
+	CookiePath() string
+	// Return the flag value used for the csrf token.
+	CookieHttpOnly() bool
+	// Return the token.
+	Token() string
+	// Validate by token.
+	ValidToken(t string) bool
+	// Error replies to the request with a custom function when ValidToken fails.
+	Error(w http.ResponseWriter)
```
Comment (unknwon) [inline csrf.go:36]:
```suggestion
	// HeaderName returns the HTTP header used to search for token.
	HeaderName() string
	// FormName returns the form value used to search for token.
	FormName() string
	// CookieName returns the cookie name used to search for token.
	CookieName() string
	// CookiePath returns the cookie path.
	CookiePath() string
	// CookieHttpOnly returns the cookie flag value used for the CSRF token.
	CookieHttpOnly() bool
	// Token returns the token.
	Token() string
	// ValidToken returns true if the token passed validation.
	ValidToken(t string) bool
	// Error replies to the request with a custom function when ValidToken fails.
	Error(w http.ResponseWriter)
```

File: csrf.go [hunk 2]
```diff
@@ -1 +1,295 @@
...
+	// Cookie path
+	cookiePath string
+	// Cookie HttpOnly flag value used for the csrf token.
+	cookieHttpOnly bool
+	// Token generated to pass via header, cookie, or hidden form value.
+	token string
+	// This value must be unique per user.
+	id string
+	// Secret used along with the unique id above to generate the Token.
+	secret string
```
Comment (unknwon) [inline csrf.go:57]:
```suggestion
	// Header name value for setting and getting CSRF token.
	header string
	// Form name value for setting and getting CSRF token.
	form string
	// Cookie name value for setting and getting CSRF token.
	cookie string
	// Cookie domain
	cookieDomain string
	// Cookie path
	cookiePath string
	// Cookie HttpOnly flag value used for the CSRF token.
	cookieHttpOnly bool
	// Token generated to pass via header, cookie, or hidden form value.
	token string
	// This value must be unique per user.
	id string
	// Secret used along with the unique id above to generate the Token.
	secret string
```

File: csrf.go [hunk 3]
```diff
@@ -1 +1,267 @@
...
+)
+
+// CSRF represents a CSRF service and is used to get the current token and validate a suspect token.
+type CSRF interface {
+	// Token returns the token.
+	Token() string
+	// ValidToken returns true if the token passed validation.
+	ValidToken(t string) bool
+	// Error replies to the request with a custom function when ValidToken fails.
+	Error(w http.ResponseWriter)
```
Comment (unknwon) [inline csrf.go:26]:
```suggestion
	// Error executes the error function with given http.ResponseWriter.
	Error(w http.ResponseWriter)
```

File: csrf.go [hunk 4]
```diff
@@ -1 +1,267 @@
...
+	// Validate validate CSRF by context.
+	// It attempts to get a token from a "X-CSRFToken"
+	// HTTP header and then a "_csrf" form value. If one of these is found, the token will be validated
+	// using ValidToken. If this validation fails, custom Error is sent in the reply.
+	// If neither a header or form value is found, http.StatusBadRequest is sent.
+	Validate(ctx flamego.Context)
```
Comment (unknwon) [inline csrf.go:32]:
```suggestion
	// Validate validates CSRF using given context.
	// It first attempts to get the token from the HTTP header ("X-CSRFToken" by default)
	// and then the form value ("_csrf" by default). If one of these is found, the token will be validated
	// using ValidToken. If this validation fails, custom Error is sent as the response.
	// If neither the header nor form value is found, http.StatusBadRequest is sent.
	Validate(ctx flamego.Context)
```

File: csrf.go [hunk 5]
```diff
@@ -1 +1,267 @@
...
+	// Header name value for setting and getting CSRF token.
+	header string
+	// Form name value for setting and getting CSRF token.
+	form string
+	// Cookie name value for setting and getting CSRF token.
+	cookie string
+	// Cookie domain
+	cookieDomain string
+	// Cookie path
+	cookiePath string
```
Comment (unknwon) [inline csrf.go:45]:
```suggestion
	// Cookie domain for setting the cookie.
	cookieDomain string
	// Cookie path for setting the cookie.
	cookiePath string
```

File: csrf.go [hunk 6]
```diff
@@ -1 +1,267 @@
...
+	// This value must be unique per user.
+	id string
+	// Secret used along with the unique id above to generate the Token.
+	secret string
+	// ErrorFunc is the custom function that replies to the request when ValidToken fails.
+	errorFunc func(w http.ResponseWriter)
+}
+
+// Token returns the current token. This is typically used
+// to populate a hidden form in an HTML template.
```
Comment (unknwon) [inline csrf.go:59]:
```suggestion
```

File: csrf.go [hunk 7]
```diff
@@ -1 +1,267 @@
...
+	"time"
+
+	"github.com/flamego/flamego"
+	"github.com/flamego/session"
+)
+
+// CSRF represents a CSRF service and is used to get the current token and validate a suspect token.
+type CSRF interface {
+	// Token returns the token.
+	Token() string
```
Comment (unknwon) [inline csrf.go:22]:
```suggestion
	// Token returns the current token. This is typically used
	// to populate a hidden form in an HTML template.
	Token() string
```

File: csrf.go [hunk 8]
```diff
@@ -1 +1,267 @@
...
+	errorFunc func(w http.ResponseWriter)
+}
+
+// Token returns the current token. This is typically used
+// to populate a hidden form in an HTML template.
+func (c *csrf) Token() string {
+	return c.token
+}
+
+// ValidToken validates the passed token against the existing Secret and ID.
```
Comment (unknwon) [inline csrf.go:64]:
```suggestion
```

File: csrf.go [hunk 9]
```diff
@@ -1 +1,267 @@
...
+func (c *csrf) ValidToken(t string) bool {
+	return ValidToken(t, c.secret, c.id, "POST")
+}
+
+// Error replies to the request when ValidToken fails.
```
Comment (unknwon) [inline csrf.go:69]:
```suggestion
```

File: csrf.go [hunk 10]
```diff
@@ -1 +1,267 @@
...
+	"github.com/flamego/flamego"
+	"github.com/flamego/session"
+)
+
+// CSRF represents a CSRF service and is used to get the current token and validate a suspect token.
+type CSRF interface {
+	// Token returns the token.
+	Token() string
+	// ValidToken returns true if the token passed validation.
+	ValidToken(t string) bool
```
Comment (unknwon) [inline csrf.go:24]:
```suggestion
	// ValidToken validates the passed token against the existing Secret and ID.
	ValidToken(t string) bool
```

File: csrf.go [hunk 11]
```diff
@@ -1 +1,267 @@
...
+	// HTTP header used to set and get token.
+	Header string
+	// Form value used to set and get token.
+	Form string
+	// Cookie value used to set and get token.
+	Cookie string
+	// Cookie domain.
+	CookieDomain string
+	// Cookie path.
+	CookiePath string
```
Comment (unknwon) [inline csrf.go:119]:
```suggestion
	// Cookie domain used to set cookie.
	CookieDomain string
	// Cookie path used to set cookie.
	CookiePath string
```

File: csrf.go [hunk 12]
```diff
@@ -1 +1,267 @@
...
+	// Enable cookie HttpOnly attribute.
+	CookieHttpOnly bool
+	// Key used for getting the unique ID per user.
+	SessionKey string
+	// oldSeesionKey saves old value corresponding to SessionKey.
+	oldSeesionKey string
```
Comment (unknwon) [inline csrf.go:125]:
```suggestion
```

Seems not useful at all, we can remove it.

File: csrf.go [hunk 13]
```diff
@@ -1 +1,243 @@
...
+		return
+	}
+
+	http.Error(ctx.ResponseWriter(), "Bad Request: no CSRF token present", http.StatusBadRequest)
+}
+
+// Options maintains options to manage behavior of Generate.
+type Options struct {
+	// The global secret value used to generate Tokens.
+	Secret string
```
Comment (unknwon) [inline csrf.go:106]:
```suggestion
	// The global secret value used to generate tokens.
	Secret string
```

File: csrf.go [hunk 14]
```diff
@@ -1 +1,243 @@
...
+	// Cookie path used to set cookie.
+	CookiePath string
+	// Enable cookie HttpOnly attribute.
+	CookieHttpOnly bool
+	// Key used for getting the unique ID per user.
+	SessionKey string
+	// If true, send token via X-CSRFToken header.
+	SetHeader bool
+	// If true, send token via _csrf cookie.
+	SetCookie bool
```
Comment (unknwon) [inline csrf.go:124]:
```suggestion
	// If true, send token via HTTP header.
	SetHeader bool
	// If true, send token via cookie.
	SetCookie bool
```

**Review by unknwon — APPROVED** [review]

File: csrf.go [hunk 15]
```diff
@@ -1 +1,192 @@
...
+		uid := s.Get(opt.SessionKey)
+		if uid != nil {
+			x.id = fmt.Sprintf("%s", uid)
+		} else {
+			x.id = "0"
+		}
+
+		x.token = GenerateToken(x.secret, x.id, http.MethodPost)
+		if opt.SetHeader {
+			c.ResponseWriter().Header().Add(opt.Header, x.token)
```
Comment (wuhan005) [inline csrf.go]:
```suggestion
			c.ResponseWriter().Header().Set(opt.Header, x.token)
```

**Review by wuhan005 — APPROVED** [review]

**Pull request merged**